src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/TypeClass.java

```diff
@@ -42,11 +42,10 @@
     private static final int MAX_AGGREGATE_REGS_SIZE = 2;
 
     private static TypeClass classifyValueType(ValueLayout type) {
-        if (!(type instanceof CLinker.CValueLayout)) {
-            throw new IllegalStateException("Unexpected value layout: could not determine ABI class");
-        }
+        CLinker.TypeKind kind = (CLinker.TypeKind)type.attribute(CLinker.TypeKind.ATTR_NAME).orElseThrow(
+                () -> { throw new IllegalStateException("Unexpected value layout: could not determine ABI class"); });
```

Comment:
Note that orElseThrow takes a `Supplier<Throwable>` (basically), so this can be simplified to return the new exception instance instead of throwing it.

Also, this code is copied in the other TypeClass classes, so might be good to create a helper method in SharedUtils for it instead?

Reply:
Done